diff --git a/mpp/hal/CMakeLists.txt b/mpp/hal/CMakeLists.txt
index 6c32664..0749f81 100644
--- a/mpp/hal/CMakeLists.txt
+++ b/mpp/hal/CMakeLists.txt
@@ -40,0 +41 @@ target_link_libraries(mpp_hal
+                      hal_common
diff --git a/mpp/hal/common/CMakeLists.txt b/mpp/hal/common/CMakeLists.txt
index addb9ae..fa83fbf 100644
--- a/mpp/hal/common/CMakeLists.txt
+++ b/mpp/hal/common/CMakeLists.txt
@@ -15,4 +15,12 @@ set_target_properties(hal_common PROPERTIES FOLDER "mpp/hal/common")
-add_subdirectory(h264)
-add_subdirectory(h265)
-add_subdirectory(jpeg)
-add_subdirectory(av1)
+if(HAVE_H264E)
+    add_subdirectory(h264)
+endif()
+if(HAVE_H265E)
+    add_subdirectory(h265)
+endif()
+if(HAVE_JPEGE)
+    add_subdirectory(jpeg)
+endif()
+if(HAVE_AV1D)
+    add_subdirectory(av1)
+endif()
diff --git a/mpp/hal/rkenc/CMakeLists.txt b/mpp/hal/rkenc/CMakeLists.txt
index cbb3d6c..9c1a2c7 100644
--- a/mpp/hal/rkenc/CMakeLists.txt
+++ b/mpp/hal/rkenc/CMakeLists.txt
@@ -2,4 +2,12 @@
-add_subdirectory(common)
-add_subdirectory(h264e)
-add_subdirectory(h265e)
-add_subdirectory(jpege)
+if(HAVE_H264E OR HAVE_H265E OR HAVE_JPEGE)
+    add_subdirectory(common)
+endif()
+if(HAVE_H264E)
+    add_subdirectory(h264e)
+endif()
+if(HAVE_H265E)
+    add_subdirectory(h265e)
+endif()
+if(HAVE_JPEGE)
+    add_subdirectory(jpege)
+endif()
diff --git a/osal/inc/mpp_hash.h b/osal/inc/mpp_hash.h
index 8cfdf64..a537147 100644
--- a/osal/inc/mpp_hash.h
+++ b/osal/inc/mpp_hash.h
@@ -46 +46 @@ struct hlist_head {
-    (*((volatile typeof(val) *)(&(var))) = (val))
+    (*((volatile __typeof__(val) *)(&(var))) = (val))
@@ -48 +48 @@ struct hlist_head {
-#define READ_ONCE(var) (*((volatile typeof(var) *)(&(var))))
+#define READ_ONCE(var) (*((volatile __typeof__(var) *)(&(var))))
@@ -154 +154 @@ static inline void hlist_move_list(struct hlist_head *old,
-    ({ typeof(ptr) ____ptr = (ptr); \
+    ({ __typeof__(ptr) ____ptr = (ptr); \
@@ -159 +159 @@ static inline void hlist_move_list(struct hlist_head *old,
-    for (pos = hlist_entry_safe((head)->first, typeof(*(pos)), member);\
+    for (pos = hlist_entry_safe((head)->first, __typeof__(*(pos)), member);\
@@ -161 +161 @@ static inline void hlist_move_list(struct hlist_head *old,
-         pos = hlist_entry_safe((pos)->member.next, typeof(*(pos)), member))
+         pos = hlist_entry_safe((pos)->member.next, __typeof__(*(pos)), member))
@@ -164 +164 @@ static inline void hlist_move_list(struct hlist_head *old,
-    for (pos = hlist_entry_safe((pos)->member.next, typeof(*(pos)), member);\
+    for (pos = hlist_entry_safe((pos)->member.next, __typeof__(*(pos)), member);\
@@ -166 +166 @@ static inline void hlist_move_list(struct hlist_head *old,
-         pos = hlist_entry_safe((pos)->member.next, typeof(*(pos)), member))
+         pos = hlist_entry_safe((pos)->member.next, __typeof__(*(pos)), member))
@@ -170 +170 @@ static inline void hlist_move_list(struct hlist_head *old,
-         pos = hlist_entry_safe((pos)->member.next, typeof(*(pos)), member))
+         pos = hlist_entry_safe((pos)->member.next, __typeof__(*(pos)), member))
@@ -173 +173 @@ static inline void hlist_move_list(struct hlist_head *old,
-    for (pos = hlist_entry_safe((head)->first, typeof(*pos), member);\
+    for (pos = hlist_entry_safe((head)->first, __typeof__(*pos), member);\
@@ -175 +175 @@ static inline void hlist_move_list(struct hlist_head *old,
-         pos = hlist_entry_safe(n, typeof(*pos), member))
+         pos = hlist_entry_safe(n, __typeof__(*pos), member))
